wp-content/civi-extensions/civirazorpay/cli/update-recurring-contact-info.php

```diff
@@ -0,0 +1,309 @@
+  }
+
+  /**
+   * Fetch subscriptions from Razorpay API.
+   *
+   * @return array
+   */
+  private function fetchSubscriptions($limit): array {
+    $options = [
+      'count' => $limit,
+      'skip' => $this->skip,
+    // Only fetch active subscriptions.
+      'status' => 'active',
+    ];
+
+    $response = $this->api->subscription->all($options);
+    $responseArray = $response->toArray();
+
+    return $responseArray['items'] ?? [];
+  }
+
+  /**
+   * Handle an individual Razorpay subscription.
+   *
+   * @param array $subscription
+   */
+  private function processSubscription(array $subscription): void {
+    echo "Processing Subscription ID: {$subscription['id']}\n";
+    echo "Status: {$subscription['status']}\n";
+    echo "Subscription ID: {$subscription['id']}\n";
+
+    try {
+      $contactID = $this->handleCustomerDataAndFindContact($subscription);
+
+      if (!$contactID) {
+        throw new Exception("No valid contact could be associated with subscription {$subscription['id']}");
+      }
+
+    }
+    catch (Exception $e) {
+      echo "Error processing subscription {$subscription['id']}: " . $e->getMessage() . "\n";
+    }
+  }
+
+  /**
+   * Handle customer data associated with the subscription.
+   *
+   * @param array $subscription
+   */
+  private function handleCustomerDataAndFindContact(array $subscription) {
+    $subscriptionId = $subscription['id'] ?? NULL;
+    $notes = $subscription['notes'] ?? [];
+    $email = $notes['email'] ?? NULL;
+    $name = $notes['name'] ?? NULL;
+    $mobile = $notes['mobile'] ?? NULL;
+    $address = $notes['address1'] ?? NULL;
+    $panCard = $notes['PAN Card'] ?? NULL;
+
+    try {
+      // Step 1: Fetch first recurring contribution to get contact ID.
+      $firstRecur = ContributionRecur::get(FALSE)
+        ->addSelect('contact_id', 'contribution.id')
+        ->addJoin('Contribution AS contribution', 'LEFT')
+        ->addWhere('processor_id', '=', $subscriptionId)
+        ->addWhere('contribution.source', '=', 'Imported from Razorpay')
+        ->execute()
+        ->first();
+
+      $contactId = $firstRecur['contact_id'] ?? NULL;
+
+      if ($contactId) {
+        echo "Contact found successfully. Contact ID: $contactId\n";
+
+        try {
+          $this->updateDetailsOnContact($contactId, $mobile, $address, $panCard);
+        }
+        catch (\Exception $e) {
+          \Civi::log()->error('Error updating contact details: ' . $e->getMessage());
+        }
+
+        // Step 2: Now fetch all contributions and update PAN card.
+        try {
+          $allRecurs = ContributionRecur::get(FALSE)
+            ->addSelect('contact_id', 'contribution.id')
+            ->addJoin('Contribution AS contribution', 'LEFT')
+            ->addWhere('processor_id', '=', $subscriptionId)
+            ->addWhere('contribution.source', '=', 'Imported from Razorpay')
+            ->execute();
```

Comment:
This looks overcomplicated to me:

```suggestion
         $contributions = \Civi\Api4\Contribution::get(FALSE)
            ->addSelect('contact_id', 'id')
            ->addWhere('contribution_recur_id', '=', $firstRecur['id'])
            ->addWhere('is_test', '=', $this->is_test)
            ->addWhere('source', '=', 'Imported from Razorpay')
            ->execute();
```

Reply:
It’s a simple and quick solution. The subscription ID is unique across all recurring subscriptions, so I fetched it from Razorpay and used it to identify the corresponding contact_id and contribution id attach to one recurring id in CiviCRM. Since the subscription ID is unique, it gave me the correct contact and contribution every time. CiviCRM  provides the ability to fetch data using ContributionRecur, so I used that to retrieve the necessary details.

I expect the approach you mentioned might not fetch all contribution IDs for the recurring contributions, since we’re applying the condition (First()) on the $firstRecur query so it fetched only top most not the other subsiquent

Reply:
resolving based on point discussed in call